doctr/datasets/svt.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (C) 2021, Mindee.
+
+# This program is licensed under the Apache License version 2.
+# See LICENSE or go to <https://www.apache.org/licenses/LICENSE-2.0.txt> for full license details.
+
+import os
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import defusedxml.ElementTree as ET
+import numpy as np
+
+from .datasets import VisionDataset
+
+__all__ = ['SVT']
+
+
+class SVT(VisionDataset):
+    """SVT dataset from `"The Street View Text Dataset - UCSD Computer Vision"
+    <http://vision.ucsd.edu/~kai/svt/>`_.
+
+    Example::
+        >>> from doctr.datasets import SVT
+        >>> train_set = SVT(train=True, download=True)
+        >>> img, target = train_set[0]
+
+    Args:
+        train: whether the subset should be the training one
+        sample_transforms: composable transformations that will be applied to each image
+        rotated_bbox: whether polygons should be considered as rotated bounding box (instead of straight ones)
+        **kwargs: keyword arguments from `VisionDataset`.
+    """
+
+    URL = 'http://vision.ucsd.edu/~kai/svt/svt.zip'
+    SHA256 = '63b3d55e6b6d1e036e2a844a20c034fe3af3c32e4d914d6e0c4a3cd43df3bebf'
+
+    def __init__(
+        self,
+        train: bool = True,
+        sample_transforms: Optional[Callable[[Any], Any]] = None,
+        rotated_bbox: bool = False,
+        **kwargs: Any,
+    ) -> None:
+
+        super().__init__(self.URL, None, self.SHA256, True, **kwargs)
+        self.sample_transforms = sample_transforms
+        self.train = train
+        self.data: List[Tuple[Path, Dict[str, Any]]] = []
+        np_dtype = np.float16 if self.fp16 else np.float32
+
+        # Load xml data
+        tmp_root = os.path.join(self.root, 'svt1')
+        xml_tree = ET.parse(os.path.join(tmp_root, 'train.xml')) if self.train else ET.parse(
+            os.path.join(tmp_root, 'test.xml'))
+        xml_root = xml_tree.getroot()
+
+        for child in xml_root:
+            for image_tag in child:
+                if image_tag.tag == 'imageName':
+                    _raw_path = str(image_tag.text)
+
+                    # File existence check
+                    if not os.path.exists(os.path.join(tmp_root, _raw_path)):
+                        raise FileNotFoundError(f"unable to locate {os.path.join(tmp_root, _raw_path)}")
+
+                if rotated_bbox:
+                    # x_center, y_center, w, h, 0
+                    _tmp_box_targets = [
+                        (int(rect_tag.attrib['x']) + int(rect_tag.attrib['width']) / 2,
+                         int(rect_tag.attrib['y']) + int(rect_tag.attrib['height']) / 2,
+                         int(rect_tag.attrib['width']), int(rect_tag.attrib['height']), 0)
+                        for rect_tag in image_tag
```

Comment:
according to what you said above, I don't see why we don't need to check that `if image_tag.tag == 'taggedRectangles'` then :thinking:

Reply:
```
image.tag : imageName
image.attrib : {}
image.text : img/04_16.jpg
image.tag : address
image.attrib : {}
image.text : 324 South Glenoaks Boulevard Burbank CA 91502-1318
image.tag : lex
image.attrib : {}
image.text : BLOCKBUSTER,VIDEO,LAW,OFFICES,RAY,FOUNTAIN,CRIMINAL,DEFENSE,ATTORNEY,LAWYER,PIZZA,MAN,CAFE,COLOMBIA,GERRO,MARSHA,RICO,GLENOAKS,PHARMACY,AND,MEDICAL,SUPPLY,RAMIREZ,LYNN,TER,POGHOSYAN,ZARINE,JAS,GIFTS,KNOX,TIMOTHY,DDS,LILLY,PARK,HOME,THEATER,HAMOUI,MOBIL,MART,CLEANERS,TOKYO,YAKIDORI,SHIRVAN,REALTY,GROUP,NAPA,AUTO,CARE,CENTER,TRAN,DEBBY
image.tag : Resolution
image.attrib : {'x': '1280', 'y': '1024'}
image.text : None
image.tag : taggedRectangles
image.attrib : {}
image.text : 
```

This is for one example hopefully it makes it a bit clearer   

Why we need no check ?
taggedRectangles has only a attribut
and tag with the label is child from taggedRectangles
but in image_tag we have multible tags: imageName, address, lex, Resolution and taggedRectangles

```
<taggedRectangle height="31" width="120" x="351" y="455">
<tag>MAGIC</tag>
</taggedRectangle>
```

Reply:
Sorry, but I'm not really familiar with XML, so I have no clue in your snippet whether there is some hierarchy, whether this is a flat data structure or whether we can access attribute values by name :sweat_smile: 

Generally speaking, nested FOR loops have to be avoided, hence my previous question ;)
So, in Python, when processing the XML, for what I understand:
- `xml_root` holds the list of pages/images
- in each page/image, there are several attributes, and we're interested in "imageName", "x", "y", "width", "height" and those are on the same level

Is this correct?

Reply:
![j](https://user-images.githubusercontent.com/23098070/141478823-c0dfffcf-ef78-4c3b-bc1d-cc52e7c3ac55.png)
No lets try to explain with this snippet:
tagset is root
image is first depth tag
imageName, address, lex, Resolution and taggedRectangles is second depth (x,y from Resolution are attributes and the other are text)
taggedRectangles has 1 more depth taggedRectangle with attributes height, width, x, y
and each taggedRectangle has one more depth tag with a text value
@fg-mindee 
i have also tryed pythons built-in import xmltodict but this does not work in fact that the xml file is not well formated :/
the other way (and i think the only other way) with XPath something like this: .//imageName  but for readability i would prefer the actual implementation :sweat_smile: 
wdyt ?